File: app_codebase/python_base_05/bin/core/ws/contracts/ws_message_contract.py
"""WebSocket message shapes — shared wire format matching HTTP JSON envelope."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class WsConnectionContext:
    """Per-connection state after tier auth (if required)."""

    tier: str
    connection_id: str
    user_id: str | None = None
    claims: dict[str, Any] = field(default_factory=dict)
    authenticated: bool = False


@dataclass
class WsClientMessage:
    """Parsed client frame body (the ``data`` object when ok)."""

    msg_type: str
    channel: str
    payload: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_data(cls, data: dict[str, Any]) -> WsClientMessage | None:
        msg_type = str(data.get("type", "")).strip()
        channel = str(data.get("channel", "")).strip()
        if not msg_type or not channel:
            return None
        payload = data.get("payload")
        if not isinstance(payload, dict):
            payload = {}
        return cls(msg_type=msg_type, channel=channel, payload=payload)

File: app_codebase/dart_bkend_base_02/bin/core/ws/contracts/ws_message_contract.dart
/// WebSocket message shapes — shared wire format matching HTTP JSON envelope.
library;

class WsConnectionContext {
  WsConnectionContext({
    required this.tier,
    required this.connectionId,
    this.userId,
    this.claims = const {},
    this.authenticated = false,
  });

  final String tier;
  final String connectionId;
  String? userId;
  Map<String, dynamic> claims;
  bool authenticated;
}

class WsClientMessage {
  WsClientMessage({
    required this.msgType,
    required this.channel,
    this.payload = const {},
  });

  final String msgType;
  final String channel;
  final Map<String, dynamic> payload;

  static WsClientMessage? fromData(Map<String, dynamic> data) {
    final msgType = data['type']?.toString().trim() ?? '';
    final channel = data['channel']?.toString().trim() ?? '';
    if (msgType.isEmpty || channel.isEmpty) return null;
    final rawPayload = data['payload'];
    final payload = rawPayload is Map<String, dynamic>
        ? rawPayload
        : rawPayload is Map
            ? Map<String, dynamic>.from(rawPayload)
            : <String, dynamic>{};
    return WsClientMessage(
      msgType: msgType,
      channel: channel,
      payload: payload,
    );
  }
}

/// Handler returns response data map, error map with `_wsSendError`, or null.
typedef WsChannelHandler = Map<String, dynamic>? Function(
  WsConnectionContext ctx,
  WsClientMessage msg,
);
